kernel/src/transaction.rs

```diff
@@ -1,120 +1,260 @@
+                (Box::new(std::iter::once(actions)), schema)
+            }
+            None => {
+                (
+                    // if there is no commit info, actions are empty iterator and schema is just our
+                    // known log schema.
+                    Box::new(std::iter::empty()),
+                    get_log_schema().clone().into(),
+                )
+            }
+        };
 
-        Ok(visitor.transactions)
+        // step two: set new commit version (current_version + 1) and path to write
+        let commit_version = self.read_snapshot.version() + 1;
+        let commit_path =
+            ParsedLogPath::new_commit(self.read_snapshot.table_root(), commit_version)?;
+
+        // step three: commit the actions as a json file in the log
+        let json_handler = engine.get_json_handler();
+        match json_handler.write_json_file(&commit_path.location, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(crate::error::Error::ObjectStore(object_store::Error::AlreadyExists {
+                ..
+            })) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
     }
 
-    /// Scan the Delta Log for the latest transaction entry of an application
-    pub fn application_transaction(
-        &self,
-        engine: &dyn Engine,
-        application_id: &str,
-    ) -> DeltaResult<Option<Transaction>> {
-        let mut transactions = self.scan_application_transactions(engine, Some(application_id))?;
-        Ok(transactions.remove(application_id))
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. Note it is required in order to commit. If the engine does not
+    /// require any commit info, pass an empty `EngineData`.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>, schema: Schema) {
+        self.commit_info = Some(EngineCommitInfo {
+            data: commit_info.into(),
+            schema: schema.into(),
+        });
     }
+}
+
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(crate::Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, crate::Version),
+}
+
+// given the engine's commit info (data and schema as [EngineCommitInfo]) we want to create both:
+// (1) the commitInfo action to commit (and append more actions to) and
+// (2) the schema for the actions to commit (this is determined by the engine's commit info schema)
+fn generate_commit_info(
```

Comment:
We do a _whole lot_ of breaking apart and putting together of schemas here. But _really_ all we want is "the existing schema, but put any extra columns specified by the engine into the `CommitInfo` schema.  It's not functional, but I think you could do this _much_ more easily via something like:

```rust

    let mut action_schema = get_log_schema().clone();
    let commit_info_field = action_schema.fields.get_mut(COMMIT_INFO_NAME).ok_or_else(|| {
        crate::Error::missing_column(COMMIT_INFO_NAME)
    })?;
    let DataType::Struct(mut commit_info_data_type) = commit_info_field.data_type().clone() else {
        // TODO: Return an error
        unreachable!("commit_info_field is a struct");
    };
    commit_info_data_type.fields.extend(engine_commit_info.schema.fields().map(|f| (f.name.clone(), f.clone())));
    commit_info_field.data_type = DataType::Struct(commit_info_data_type);
```

Reply:
ah very nice thanks for the help here! had gone back and forth a bit on functional/imperative but this is nice heh